File: examples/bookShopPaper.erl
-module(bookShopPaper).
-export([main/0]).

main() -> spawn(fun() -> f() end).

f() ->
  register(p, self()),
  receive 
    _ -> loop([])
  end.

recvAnything() ->
  receive
    V -> V
  end.

loop(Books) ->
  receive
    {Hs, "buy"} ->
      B = recvAnything(),
      Books2 =
      if
        B >= 0 -> [ B || Books ];
        true   -> Books
      end,
      loop(Books2);
    {Hs, "request"} -> Hs ! Books,
                       loop(Books);
    {Hs, "exit"} -> exit(done)
  end.

File: examples/BookShopPaper.hs
{-# LANGUAGE TypeOperators
           , FlexibleContexts #-}
module BookShop where

import SessionCheck
import SessionCheck.Backend.Erlang

bookShop :: (Int :< t, String :< t, [Int] :< t) => Spec t ()
bookShop = bookShopLoop []

bookShopLoop :: (Int :< t, String :< t, [Int] :< t) => [Int] -> Spec t ()
bookShopLoop books = do
  choice <- choose [ "buy", "request", "exit" ]
  case choice of
    "buy" -> do
      book <- send anyInt
      bookShopLoop (book : books)
    "request" -> do
      get (permutationOf books)
      bookShopLoop books
    "exit" -> do
      stop

main :: IO ()
main = do
  erlangMain "bookShopPaper:main" bookShop
